File: app/actors/DashboardActor.java
package actors;

import java.util.HashSet;

import javax.inject.Inject;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.google.inject.assistedinject.Assisted;

import akka.actor.AbstractActor;
import akka.actor.Actor;
import akka.actor.ActorRef;
import akka.japi.pf.ReceiveBuilder;
import akka.persistence.PersistentActor;
import pojo.Dashboard;
import pojo.Item;

/**
 * Created by vasek on 11. 12. 2016.
 */
public class DashboardActor extends PersistentActor {

    private static final Logger LOG = LoggerFactory.getLogger(DashboardActor.class);

    private Dashboard dashboard;

    private final HashSet<ActorRef> watchers = new HashSet<>();

    @Inject
    public DashboardActor(@Assisted("dashboard") Dashboard dashboard) {
        this.dashboard = dashboard;
    }

    
    private void handleCommand(Dashboard.IncrementItem command) {
        Item item = dashboard.getItems().get(command.name);
        if (item != null) {
            item.increment();
        }
        notifyWatchers();
    }

    private void handleCommand(Dashboard.DecrementItem command) {
        Item item = dashboard.getItems().get(command.name);
        if (item != null) {
            item.decrement();
        }
        notifyWatchers();
    }

    private void handleCommand(Dashboard.Watch command) {
        Dashboard.Data data = new Dashboard.Data(dashboard.getItems());
        getSender().tell(data, self());
        watchers.add(getSender());
    }

    private void handleCommand(Dashboard.Unwatch command) {
        watchers.remove(getSender());
    }

    private void handleCommand(Dashboard.GetWriteHash command) {
        getSender().tell(dashboard.getWriteHash(), self());
    }

    private void handleCommand(Dashboard.GetReadonlyHash command) {
        getSender().tell(dashboard.getReadOnlyHash(), self());
    }

    private void handleCommand(Dashboard.GetName command) {
        getSender().tell(dashboard.getName(), self());
    }

    private void handleCommand(Dashboard.AddItem command) {
        Dashboard.AddItem addItem = (Dashboard.AddItem)command;
        dashboard.getItems().put(addItem.name, new Item(addItem.name));

        notifyWatchers();
    }

    private void handleCommand(Dashboard.RemoveItem command) {
        Dashboard.RemoveItem removeItem = (Dashboard.RemoveItem)command;
        dashboard.getItems().remove(removeItem.name);

        notifyWatchers();
    }


    //@Override
    public void onReceive(Object message) throws Throwable {


//        if (message instanceof Dashboard.IncrementItem) {
//            Dashboard.IncrementItem incrementItem = (Dashboard.IncrementItem)message;
//            Item item = dashboard.getItems().get(incrementItem.name);
//            if (item != null) {
//                item.increment();
//            }
//            notifyWatchers();
//        }
//
//        if (message instanceof Dashboard.DecrementItem) {
//            Dashboard.DecrementItem decrementItem = (Dashboard.DecrementItem)message;
//            Item item = dashboard.getItems().get(decrementItem.name);
//            if (item != null) {
//                item.decrement();
//            }
//            notifyWatchers();
//        }
//
//        if (message instanceof Dashboard.Watch) {
//            Dashboard.Data data = new Dashboard.Data(dashboard.getItems());
//            getSender().tell(data, self());
//            watchers.add(getSender());
//        }

//        if (message instanceof Dashboard.Unwatch) {
//            watchers.remove(getSender());
//        }

//        if (message instanceof Dashboard.GetWriteHash) {
//            getSender().tell(dashboard.getWriteHash(), self());
//        }

//        if (message instanceof Dashboard.GetReadonlyHash) {
//            getSender().tell(dashboard.getReadOnlyHash(), self());
//        }
//
//
//        if (message instanceof Dashboard.GetName) {
//            getSender().tell(dashboard.getName(), self());
//        }
//
//        if (message instanceof Dashboard.AddItem) {
//            Dashboard.AddItem addItem = (Dashboard.AddItem)message;
//            dashboard.getItems().put(addItem.name, new Item(addItem.name));
//
//            notifyWatchers();
//
//        }
//
//        if (message instanceof Dashboard.RemoveItem) {
//            Dashboard.RemoveItem removeItem = (Dashboard.RemoveItem)message;
//            dashboard.getItems().remove(removeItem.name);
//
//            notifyWatchers();
//
//        }



//        ReceiveBuilder
//                .match(Dashboard.Item.class, latest -> {
//
//                    // add a new stock price to the history and drop the oldest
//                    //Double newPrice = stockQuote.newPrice(stockHistory.peekLast());
//                    //stockHistory.add(newPrice);
//                    //stockHistory.remove();
//                    // notify watchers
//                    //watchers.forEach(watcher -> watcher.tell(new Stock.Update(symbol, newPrice), self()));
//
//                    // add new name on showDashboard
//
//                })
//                .match(Stock.Watch.class, watch -> {
//                    // reply with the stock history, and add the sender as a watcher
//                    //final Double[] clone = stockHistory.toArray(new Double[]{});
//                    //sender().tell(new Stock.History(symbol, clone), self());
//                    //watchers.add(sender());
//                })
//                .match(Stock.Unwatch.class, unwatch -> {
//                    //watchers.remove(sender());
//                    //if (watchers.isEmpty()) {
//                    //    stockTick.ifPresent(Cancellable::cancel);
//                    //    context().stop(self());
//                    //}
//                });
    }

    private void notifyWatchers() {
        watchers.forEach(watcher -> watcher.tell(new Dashboard.Data(dashboard.getItems()), self()));
    }

    protected abstract AbstractActor.Receive buildReceiver(ReceiveBuilder defaultMatches);

    private ReceiveBuilder getDefaultMatches() {
        return receiveBuilder()
                .match(Dashboard.IncrementItem.class, this::handleCommand)
                .match(Dashboard.DecrementItem.class, this::handleCommand)
                .match(Dashboard.AddItem.class, this::handleCommand)
                .match(Dashboard.RemoveItem.class, this::handleCommand)
                .match(Dashboard.Watch.class, this::handleCommand)
                .match(Dashboard.GetName.class, this::handleCommand)
                .match(Dashboard.GetReadonlyHash.class, this::handleCommand)
                .match(Dashboard.GetWriteHash.class, this::handleCommand);
    }

    @Override
    public Receive createReceiveRecover() {
        return null;
    }

    @Override
    public Receive createReceive() {
        return null;
    }

    @Override
    public String persistenceId() {
        return null;
    }

    @Override
    public void deleteSnapshot(final long sequenceNr) {
        super.deleteSnapshot(sequenceNr);
    }


//    @Override
//    public Receive createReceiveRecover() {
//        LOG.debug("receiveCommand");
//        ReceiveBuilder defaultMatches = getDefaultMatches();
//        return buildReceiver(defaultMatches);
//    }
//
//
//    @Override
//    public Receive createReceive() {
//        LOG.debug("receiveCommand");
//        ReceiveBuilder defaultMatches = getDefaultMatches();
//        return buildReceiver(defaultMatches);
//
//    }
//
//    @Override
//    public String persistenceId() {
//        return dashboard.getWriteHash();
//    }
//
//    @Override
//    public void deleteSnapshot(final long sequenceNr) {
//        super.deleteSnapshot(sequenceNr);
//    }


    public interface Factory {
        Actor create(@Assisted("name") String name, @Assisted("hash") String hash);
    }
}

File: app/actors/DashboardActor.scala
package actors

import actors.DashboardActor.Command
import actors.DashboardActor.Command.{Data, DecrementItem, IncrementItem, Watch}
import akka.actor.{ActorRef, Props}
import akka.persistence.{PersistentActor, SnapshotOffer}
import common.{EntityEvent, EntityFieldsObject, PersistentEntity}
import org.slf4j.LoggerFactory

import scala.collection.mutable

case class Item(var name: String, var score: Int) {

  def this(name: String ) {
    this(name, 0)
  }

  def increment(): Unit = this.score = this.score + 1

  def decrement(): Unit = if (this.score > 0) this.score = this.score - 1


}

case class Dashboard(id: String,
                      name: String,
                      description: String,
                      style: String,
                      items: mutable.Map[String, Item] = mutable.Map(),
                     ownerName: String,
                     ownerEmail: String,
                     readonlyHash: String,
                     writeHash: String,
                    deleted: Boolean = false
                    ) extends EntityFieldsObject[String, Dashboard] {
  /**
    * Assigns an id to the fields object, returning a new instance
    *
    * @param id The id to assign
    */
  override def assignId(id: String) = this.copy(id = id)
  override def markDeleted = this.copy(deleted = false)
}

object DashboardActor {

  private val LOG = LoggerFactory.getLogger(classOf[DashboardActor])

  val EntityType = "dashboard"

//  trait Factory {
//    def create(@Assisted("dashboard") dashboard: Dashboard): Actor
//  }

  //def props(dashboard: Dashboard): Props = Props(new DashboardActor(dashboard))

  object Command {
    case class CreateDashboard(dashboard: Dashboard)
    case class Watch()
    case class Unwatch()
    case class Data(items: mutable.Map[String, Item])
    case class IncrementItem(name: String)
    case class DecrementItem(name: String)
    case class AddItem(name: String)
    case class RemoveItem(name: String)
    case class GetWriteHash()
    case class GetReadonlyHash()
    case class GetName()
    case class GetDashboard()
  }

  object Event {
    trait DasboardEvent extends EntityEvent{override def entityType: String = EntityType}
    case class DashboardCreated(dashboard: Dashboard) extends DasboardEvent {
      def toDataModel = {
        val dashboardDM = Datamodel.Dashboard.new

      }
    }
  }

 

}

class DashboardActor(id: String) extends PersistentEntity[Dashboard](id) {

  final private val watchers: Set[ActorRef] = Set();

  private def handleIncrementItemCommand(command: IncrementItem): Unit = {
    val item: Item = dashboard.items(command.name)
    if (item != null) {
      item.increment()
    }
    notifyWatchers()
  }

  private def handleDecremenItemCommand (command: DecrementItem): Unit = {
    val item: Item = dashboard.items(command.name)
    if (item != null) {
      item.decrement ()
    }
    notifyWatchers ()
  }

  private def handleWatchCommand (command: Watch): Unit = {
    val data: Dashboard.Data = new Dashboard.Data(dashboard.items)
    sender.tell (data, self)
    watchers.+(sender)
  }

  private def handleDataCommand (command: Data): Unit = {
    val data: Dashboard.Data = new Dashboard.Data (dashboard.items)
    sender.tell (data, self)
  }

  private def handleUnwatchCommand (command: Unwatch): Unit = {
    watchers.-(sender())
  }

  private def handleGetWriteHashCommand (command: GetWriteHash): Unit = {
    sender().tell (dashboard.writeHash, self)
  }

  private def handleGetReadonlyHashCommand (command: GetReadonlyHash): Unit = {
    sender().tell (dashboard.readonlyHash, self)
  }

  private def handleGetNameCommand (command: GetName): Unit = {
    sender().tell (dashboard.name, self)
  }

  private def handlerGetDashboard(command: GetDashboard): Unit = {
    sender().tell(dashboard, self)
  }

  private def handleAddItemCommand (command: AddItem): Unit = {
    val addItem: Dashboard.AddItem = command.asInstanceOf[Dashboard.AddItem]
    dashboard.items + addItem.name
    notifyWatchers ()
  }

  private def handleRemoveItemCommand (command: RemoveItem): Unit = {
    val removeItem: Dashboard.RemoveItem = command.asInstanceOf[Dashboard.RemoveItem]
    dashboard.items - removeItem.name
    notifyWatchers ()
  }

  private def notifyWatchers (): Unit = {
    for (watcher:ActorRef <- watchers) {
      watcher.tell(new Dashboard.Data (dashboard.items), self);
    }
  }

  override def receiveRecover: Receive = {
    case SnapshotOffer(_, snapshot: Dashboard) =>
      dashboard = snapshot
  }

  override def receiveCommand: Receive = {
    case cmd:Dashboard.RemoveItem => handleRemoveItemCommand(cmd)
    case cmd:Dashboard.AddItem => handleAddItemCommand(cmd)
    case cmd:Dashboard.DecrementItem => handleDecremenItemCommand(cmd)
    case cmd:Dashboard.IncrementItem => handleIncrementItemCommand(cmd)
    case cmd:Dashboard.Data => handleDataCommand(cmd)
    case cmd:Dashboard.GetName => handleGetNameCommand(cmd)
    case cmd:Dashboard.GetReadonlyHash => handleGetReadonlyHashCommand(cmd)
    case cmd:Dashboard.GetWriteHash => handleGetWriteHashCommand(cmd)
    case cmd:Dashboard.GetDashboard => handlerGetDashboard(cmd)

    saveSnapshot(dashboard)
  }

  override def persistenceId: String = dashboard.writeHash
}